### PR TITLE
Rsa keygen updates

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/rsa/rsa_gen.c [hunk 1]
```diff
@@ -425,12 +425,13 @@ static int rsa_keygen(OSSL_LIB_CTX *libctx, RSA *rsa, int bits, int primes,
...
 
     /*
      * Only multi-prime keys or insecure keys with a small key length will use
      * the older rsa_multiprime_keygen().
      */
-    if (primes == 2 && bits >= 2048)
+    if (primes == 2 && bits >= 2048 && ebits > 16)
         ok = ossl_rsa_sp800_56b_generate_key(rsa, bits, e_value, cb);
 #ifndef FIPS_MODULE
     else
```
Comment (t8m) [inline crypto/rsa/rsa_gen.c:437]: I see that in FIPS_MODULE we just return an error return value, should we also raise an Error? (ERR_R_PASSED_INVALID_ARGUMENT or ERR_R_UNSUPPORTED?)

File: crypto/rsa/rsa_gen.c [hunk 2]
```diff
@@ -425,12 +425,13 @@ static int rsa_keygen(OSSL_LIB_CTX *libctx, RSA *rsa, int bits, int primes,
                       BIGNUM *e_value, BN_GENCB *cb, int pairwise_test)
 {
     int ok = 0;
+    int ebits = e_value != NULL ? BN_num_bits(e_value) : 17;
 
     /*
      * Only multi-prime keys or insecure keys with a small key length will use
```
Comment (ifranzki) [inline crypto/rsa/rsa_gen.c:431]: maybe also update the comment?
_...insecure keys with a small key length **or a public exponent <= 2^16** will use ..._

File: crypto/rsa/rsa_gen.c (hunk 1 above)
Comment (slontis) [reply crypto/rsa/rsa_gen.c:437]: Changed it to be in its own #ifdef..  ossl_rsa_sp800_56b_generate_key() checks the args in FIPS_MODE.

**Review by t8m — APPROVED** [review]
That cleaned up things.

**Review by ifranzki — APPROVED** [review]

**Review by bernd-edlinger — APPROVED** [review]
LGTM

**Comment by bernd-edlinger** [comment]
As a side-note, I wanted to mention, what I read here https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-4.pdf
at page 83:
> This algorithm requires that N1+N2≤ L – [L/2] – 4. Values for N1 and N2should be chosen such that N1+N2≤ (L/2) – log2(L) – 7, to ensure that the algorithm can generate as many as 5L distinct candidates for p.

This implies that the limits used here https://github.com/openssl/openssl/blob/1a01e5c29dfaf09af3960b4c8e6ec0f8171eda80/crypto/bn/bn_rsa_fips186_4.c#L80-L88
might make it impossible to generate as many as 20L distinct candidates for p.

**Comment by slontis** [comment]
Doesn't C10 Relate to Provable Primes.. (not probable primes?)

**Comment by bernd-edlinger** [comment]
Yeah, but this gives the clue why the values for Probable Primes in Table B1
are chosen this way.
So if p1, p2 are chosen so that len(p1) + len(p2) < 1007,
then I can say that 2*p1*p2 < 2^1008
If I check this with sage math, how many integers divisible by 2^1008 are between sqrt(2)*2^1023..2^1024
I get
```
sage: int ((2^1024-1414*2^1023/1000) / 2^1008)
19202
```
so more than 5*1024 = 5120 but less than 20*1024 = 20480.

**Comment by slontis** [comment]
There is a note at the bottom of page 34 of https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5-draft.pdf that says what the formula is..

**Comment by bernd-edlinger** [comment]
Yes, that foot note was already in the previous version, but still without any rationale.
Note that the draft you just referenced changed the limit for nlen = 2047 from "< 1007" to "<= 1007"
I always wondered why the table B1 had apparently a +/-1 error there.
Our limit is still "< 1007" but if it would be changed to "<= 1007" the number of prime candidates drops to 9597:
```
sage: int ((2^1024 - 2^1023*14142/10000) / 2^1009)
9597
```

**Comment by t8m** [comment]
Merged to master and openssl-3.0 branches. Thank you.

**Pull request closed without merging**